docker-compose-server.yml

```diff
@@ -0,0 +1,44 @@
+version: "3.7"
+
+services:
+  caddy:
+    build:
+      context: docker/caddy
+      dockerfile: Dockerfile
+    restart: unless-stopped
+    ports:
+      - "80:80"
+      - "443:443"
+    environment:
+      - SITE_ADDRESS=kosa2.pariyatti.app
+    volumes:
+      - $PWD/site:/srv
+      - caddy_data:/data
+      - caddy_config:/config
+  db:
+    image: postgres:14
+    volumes:
+      - ./volumes/db:/var/lib/postgresql/data
+    env_file:
+      - .dockerenv
+      - .env.local_password
+  web:
+    build:
+      context: .
+      dockerfile: Dockerfile
+    restart: always
+    command: bash -c "rm -f tmp/pids/server.pid && bundle exec rails s -p 3000 -b '0.0.0.0'"
+    ports:
+      - "3000:3000"
+    env_file:
+      - .dockerenv
+      - .env.local_password
+      - /home/ec2-user/.kosa/kosa-rails.dockerenv
+    environment:
+      - RUN_INGESTION=yes
+    depends_on:
+      - db
+
+volumes:
```

Comment:
So, there are two common approaches to data persistence across auto scaling groups. Both of them are chosen based on context:

1. EC2 EBS volumes: So, elastic block storage (EBS) is how data is stored for most of the EC2 VMs. There is an ability to attach multiple EBS volumes to the VM, the way we can attach multiple hard drives to a physical computer and mount them as /dev/sda1, /dev/sda2 etc. Because of the nature of the implementation of EBS volumes, they can only be attached to one instance at a time and that too to a VM in the same availability zone. i.e. if EBS volume is created in us-east-1b for example, it cannot be attached to a VM in us-east-1c. So, while EBS volumes are cheaper, there is additional logic needed to handle the attachment and mounting of EBS volumes in the VMs, especially in case of Auto Scaling Groups for EC2 VMs. In ecosystems like ECS and EKS, there are controllers in the system itself that handle this complexity. For us, to introduce ECS or EKS means significant increase in usage of AWS credits as all these modes need load balancers.

2. EFS mounts: These are provisioned NFS v4 storage volumes that can be mounted in multiple VMs. They do not need additional step of detaching a volume and attaching it to the new VM. So, a simple implementation would be that the NFS volume is mounted on the VM at a known location on startup (e.g. /efs/) and the volume for postgresql that is to be persisted across VM recycling will be stored in `/efs/volumes/db:/var/lib/postgresql/data`. Now, by design postgresql is designed to run on storage that is not shared, so running two instances of postgresql servers pointing to same underlying NFS volume storage can have unpredictable outcomes. Hence, the moment we choose to use EFS, we are making an underlying assumption that at any given point there is only one DB container that runs for a particular environment. This is a relatively safe assumption to make at this point for our single instance auto scaling group, though the moment we need to have some sort of redundancy or horizontal scaling in our setup, we have to move to alternative ways of running DB(most probably managed postgresql instance).